icu4c/source/i18n/number_skeletons.cpp

```diff
@@ -1500,32 +1500,33 @@ bool GeneratorHelpers::notation(const MacroProps& macros, UnicodeString& sb, UEr
 }
 
 bool GeneratorHelpers::unit(const MacroProps& macros, UnicodeString& sb, UErrorCode& status) {
-    if (utils::unitIsCurrency(macros.unit)) {
+    MeasureUnit unit = macros.unit;
+    if (!utils::unitIsBaseUnit(macros.perUnit)) {
+        if (utils::unitIsCurrency(macros.unit) || utils::unitIsCurrency(macros.perUnit)) {
```

Comment:
Isn't the !utils::unitIsBaseUnit(macros.perUnit) && utils::unitIsCurrency(macros.unit)) => U_UNSUPPORTED_ERROR / false
a new behavior?

Reply:
You're right that I've hereby changed the toSkeleton output. I'm not sure what the best thing to do is - is it just a matter of needing to document the behaviour change?

When constructing a NumberFormatter, what happens (in number_formatimpl.cpp, macrosToMicroGenerator FWIW) is if unit is set to currency, perUnit was simply ignored in 67, and is still ignored in 68. Is it better to preserve this behaviour, simply ignoring perUnit?

Reply:
I'll add the suggested unit tests. Thoughts on the question above ("is the change reasonable?"):

- It is indeed a behaviour change. I do think the change is reasonable, though arguably we should give an error already at the formatter construction, not only at the toSkeleton() call... and I may have given insufficient thought to what we might end up breaking with this? Do you think it's risky?

- For the remaining part of ICU-20941 for ICU 69, we *could* consider adding support for $CURRENCY-per-meter, but this doesn't fall into the unit identifier spec, so we may need to carefully consider the right way to deal with this case (e.g. also in the specs).

Reply:
We already announced that skeleton behavior for units is changing in ICU 68.  The RC version of that behavior was buggy, as Frank found.  I think it is 110% in scope to fix behavior between RC and GA according to requests from clients like Frank.

Reply:
Turns out forSkeleton doesn't support currency: blueprint_helpers::parseMeasureUnitOption returns an U_INTERNAL_PROGRAM_ERROR because there are more than 40 currencies - it supports only unit types with up to 40 subtypes.

The new toSkeleton behaviour can still be triggered by constructing a NumberFormatter with the fluent API.

Reply:
"currency-per-meter" requires more work, not only in skeletons but also in the formatting step.

I like the currently proposed fix because it fixes "percent-per-meter", for which we have formatting data, without touching compound currency units, for which we do not have formatting data.

Reply:
Changes made:
- for currencies, reject `perUnit` at `forSkeleton` time, just as `unit` is rejected.
- unit test demonstrating toSkeleton fails when fluent-API is used to construct a currency/unit conflict.

For ICU 69, we may need to revisit what is produced by the formatter: possibly an error, rather than ignoring perUnit, or possibly supporting it, after figuring out how the skeletons will be handled for currency-vs-unit.

Reply:
I have no strong opinion about CURRENCY-per-other unit as well as 
1. you add a unit test for that
2. you understand the impact of the changes 
3. V8 will not hit that case and it will not impact the v8/chrome integration work.
I will leave it to you and shane to decide is the changed behavior is acceptable to 68 or not. But we need to have Unit test to test whatever behavior you and shane agree to.